### PR TITLE
allow null region filtering

### LINKED ISSUE #3357: API unable to filter sites on region when region is null (not present)
### Environment
* Python version:  2.7.16
* NetBox version:  59139502ac7d (v2.6.1)

### Steps to Reproduce
1. Create a site without region as parent
`
"id": 2,
            "name": "Without region",
            "slug": "without-region",
            "status": {
                "value": 1,
                "label": "Active"
            },
            "region": null,
            ...
`
2. Try to get this site by site API with region filter to null
`
url -H "Accept: application/json; indent=4" -H "Authorization: Token 0123456789abcdef0123456789abcdef01234567" "http://localhost:32768/api/dcim/sites/?region=null"
`
3. Same result with region_id field filter

### Expected Behavior
Site without region must be returned in JSON format

### Observed Behavior
`
<!DOCTYPE html>
<html lang="en">

<head>
    <title>Server Error</title>
    <link rel="stylesheet" href="/static/bootstrap-3.3.7-dist/css/bootstrap.min.css">
    <link rel="stylesheet" href="/static/font-awesome-4.7.0/css/font-awesome.min.css">
    <meta charset="UTF-8">
</head>

<body>
    <div class="container-fluid">
        <div class="row">
            <div class="col-md-6 col-md-offset-3">
                <div class="panel panel-danger" style="margin-top: 200px">
                    <div class="panel-heading">
                        <strong>
                            <i class="fa fa-warning"></i>
                            Server Error
                        </strong>
                    </div>
                    <div class="panel-body">
                        
                            <p>
                                There was a problem with your request. Please contact an administrator.
                            </p>
                        
                        <hr />
                        <p>
                            The complete exception is provided below:
                        </p>
<pre><strong>&lt;class &#39;AttributeError&#39;&gt;</strong><br />
&#39;str&#39; object has no attribute &#39;get_descendants&#39;</pre>
                        <p>
                            If further assistance is required, please post to the <a href="https://groups.google.com/forum/#!forum/netbox-discuss">NetBox mailing list</a>.
                        </p>
                        <div class="text-right">
                            <a href="/" class="btn btn-primary">Home Page</a>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</body>

</html>
`

### PR DESCRIPTION
**Pull request by kobayashi**
### Fixes: #3357 

This fix allows to filter region with null value. The output is like the following.
```
$ curl -X GET -H "Accept: application/json; indent=4" -H "Authorization: Token 0123456789abcdef01234567
89abcdef01234567" "http://127.0.0.1:8000/api/dcim/sites/?region=null"

{                                                                                                   
    "count": 1,                                                                                     
    "next": null,                                                                                   
    "previous": null,                                                                               
    "results": [                                                                                    
        {                                                                                           
            "id": 2,                                                                                
            "name": "site null",                                                                    
            "slug": "site-null",
            "status": {
                "value": 1,
                "label": "Active"
            },
            "region": null,
            "tenant": null,
            "facility": "",
            "asn": null,
            "time_zone": null,
            "description": "",
            "physical_address": "",
            "shipping_address": "",
            "latitude": null,
            "longitude": null,
            "contact_name": "",                                                                     
            "contact_phone": "",
            "contact_email": "",
            "comments": "",
            "tags": [],
            "custom_fields": {},
            "created": "2019-10-04",                                                                
            "last_updated": "2019-10-04T03:50:34.076991Z",                                          
            "circuit_count": null,
            "device_count": null,
            "prefix_count": null,
            "rack_count": null,
            "virtualmachine_count": null,
            "vlan_count": null
        }
    ]
}

```

### CODE REVIEW

File: netbox/dcim/tests/test_api.py [hunk 1]
```diff
@@ -1828,6 +1839,21 @@ def test_list_devices_brief(self):
...
+            device_type=self.devicetype1,
+            device_role=self.devicerole1,
+            name='Test Device Null Region',
+            site=self.site2,
+            cluster=self.cluster1
+        )
+
+        url = reverse('dcim-api:device-list')
+        response = self.client.get('{}?region=null'.format(url), **self.header)
+
```
Comment (DanSheps) [inline netbox/dcim/tests/test_api.py]: Could we also add a test to make sure if region is not defined or is defined but empty it returns all?  Same with multiple values (region=null,12)

File: netbox/utilities/filters.py [hunk 2]
```diff
@@ -63,6 +63,11 @@ class TreeNodeMultipleChoiceFilter(django_filters.ModelMultipleChoiceFilter):
     Filters for a set of Models, including all descendant models within a Tree.  Example: [<Region: R1>,<Region: R2>]
     """
     def filter(self, qs, value):
+        if 'null' in value:
```
Comment (jeremystretch) [inline netbox/utilities/filters.py]: Convention requires that we refer to `settings.FILTERS_NULL_CHOICE_VALUE` for the null value. This is currently the string `"null"` but it could change in the future.

File: netbox/dcim/tests/test_api.py (hunk 1 above)
Comment (DanSheps) [reply netbox/dcim/tests/test_api.py]: Spoke with @kobayashi I will enhance the test suite around regions after this PR is merged.

**Comment by jeremystretch** [comment]
@kobayashi could you rebase this against the current `develop` branch? Looks like some unrelated commits got caught up in the PR.

**Comment by kobayashi** [comment]
Sorry, that was my mistake. That was rebased

File: netbox/utilities/filters.py [hunk 3]
```diff
@@ -62,7 +61,13 @@ class TreeNodeMultipleChoiceFilter(django_filters.ModelMultipleChoiceFilter):
     """
     Filters for a set of Models, including all descendant models within a Tree.  Example: [<Region: R1>,<Region: R2>]
     """
+
     def filter(self, qs, value):
+        if settings.FILTERS_NULL_CHOICE_VALUE in value:
+            # Filtering by null value. Example: region=null
+            qs = self.get_method(qs)(**{self.field_name.replace('in', 'isnull'): True})
+            return qs.distinct() if self.distinct else qs
```
Comment (jeremystretch) [inline netbox/utilities/filters.py]: This approach doesn't allow the user to specify `null` OR a specific region(s); e.g. `GET /api/dcim/sites/?region=us&region=null`. Can we extend this to modify the queryset to look like `Site.objects.filter(Q(region__isnull=True) | Q(region__in=[...]))`? Not sure how painful that is to do within the filter.

Comment (kobayashi) [reply netbox/utilities/filters.py]: Oh, I will try to find the way later. thx

Comment (kobayashi) [reply netbox/utilities/filters.py]: Applied to override `get_filter_predicate` for filtering 'null' value. This makes us possible to filter multiple region filters including 'null'. I also add tests to check this.

**Comment by jeremystretch** [comment]
Nice work on the filter! We probably don't need tests for each instance of `TreeNodeMultipleChoiceFilter` though; just filtering sites by region should be sufficient.

Long-term, we might want to establish a separate set of tests specifically for filters to keep the API tests from getting too cluttered.

**Comment by jeremystretch** [comment]
I'm going to merge this so that it gets included in v2.6.7. I'll take care of cleaning up the tests afterward.
